### PR TITLE
Performance improvements for bundle-export function

### PR DESCRIPTION
**Pull request by DeLaGuardo**
> Related https://github.com/advthreat/iroh/issues/4958

Optimizations for bundle-export functionality:

- [x] Extend `IStore` protocol with `read-records` method to fetch multiple records using a single ES search request.
- [x] Add the ability to fetch relationships related to the bunch of records using two ES requests instead of doing an ES request per id.
- [x] Replace iteration over provided ids with bulk operations in `bundle-export` logic.

## `scroll-with-limit` implementation details

The result of each invocation of `list-records` can have two main possible cases:

1. all results belong to the same record. The next call to `list-records` must use `search_after` to 'fast forward' the cursor to the next record id.
2. results list begins with records belonging to the same id, but then the tail might have 1-N partitions. That means the last partition in the result list is incomplete, and we must use `offset` to fetch the next page.

### `search_after` optimization

The second element must contain a numeric value guaranteed not to be contained in any document. For the timestamp, the value is -1. Only in this case, using "search_after" will ensure that the cursor moves to the next identifier.

<a name="qa">[§](#qa)</a> QA
============================

1. Import that bundle [bundle-0.json.zip](https://github.com/threatgrid/ctia/files/8463390/bundle-0.json.zip). It contains two sightings. Note their ids after a successful import.
2. Replace in every file from that archive [bundles.zip](https://github.com/threatgrid/ctia/files/8463400/bundles.zip) all occurrences of substrings `[[0]]` and `[[1]]` with sighting's ids from the result of step 1.
3. Import all 5 bundles.
4. Call bundle export with ids from the result of step 1. It should retrieve corresponding sightings together with related indicators. The result should not contain more than 1000 indicators per sighting.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: preffer using `ctia.store/read-records` method instead of `ctia.store/read-record` in the loop to fetch multiple records by its ids.
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

**Review by ereteog — COMMENTED** [review]
that's nice.
I need a second pass to check details, but here is a first feedback

File: src/ctia/bulk/core.clj [hunk 1]
```diff
@@ -99,16 +93,13 @@
...
-    (pmap (fn [id]
-           (try
-             (some-> (read-entity id)
-                     (with-long-id services))
-             (catch Exception e
-               (log/error (pr-str e)))))
-         ids)))
+   {{:keys [get-store]} :StoreService
+    :as services} :- ReadEntitiesServices]
+  (when-let [store (get-store entity-type)]
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bulk/core.clj:98]: I think `get-store` throws when it can't find a store for a particular entity type. And I added when-let to cover one specific test, so I will remake the test :)

File: src/ctia/bundle/core.clj [hunk 2]
```diff
@@ -321,13 +318,12 @@
   "Returns true if this entity'ID is hosted by this CTIA instance,
    false otherwise"
   [id services :- HTTPShowServices]
-  (if (seq id)
+  (when id
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj]: I guess better to revert those changes and add the test case :) anyway they are not important for that particular feature

File: src/ctia/bundle/core.clj [hunk 3]
```diff
@@ -340,155 +336,165 @@
...
+                  (assoc params-map :search_after search_after))
+           (recur (reduce into acc (cons (take limit (concat prev-data chunk)) (butlast chunks)))
+                  (last chunks) f store filter-map identity-map
+                  (assoc params-map :offset offset))))))))
+
+(defn relationships-filter [records related-to source-type target-type]
+  {:query (cond->> [(format "%s:(%s)" (name related-to) (string/join " OR " (map #(format "\"%s\"" %) (map :id records))))]
+            source-type (cons (format "source_ref:*%s*" (name source-type)))
+            target-type (cons (format "target_ref:*%s*" (name target-type)))
+            :always (reverse)
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj:405]: yes, `reverse` is here to make wildcard filter faster. I'll add a comment

File: src/ctia/bundle/core.clj [hunk 4]
```diff
@@ -340,155 +336,165 @@
...
+         (if (empty? chunks)
+           (recur (into acc (take limit (concat prev-data chunk)))
+                  [] f store filter-map identity-map
+                  (assoc params-map :search_after search_after))
+           (recur (reduce into acc (cons (take limit (concat prev-data chunk)) (butlast chunks)))
+                  (last chunks) f store filter-map identity-map
+                  (assoc params-map :offset offset))))))))
+
+(defn relationships-filter [records related-to source-type target-type]
+  {:query (cond->> [(format "%s:(%s)" (name related-to) (string/join " OR " (map #(format "\"%s\"" %) (map :id records))))]
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj:402]: I intended to gain as much control over query building as possible. To be sure that the fastest filter appears first. I can test the difference as well, but then the performance will depend on implementation details and might degrade in case of unrelated changes in the lib.

To be sure I understand you. Your suggestion is to add `:source_ref [id1 ...]` alongside with `:query`?

File: src/ctia/bundle/core.clj [hunk 5]
```diff
@@ -340,155 +336,165 @@
...
-      (send-event {:service "Export bundle fetch record"
+          store (get-store :relationship)
+          limit (get-in-config [:ctia :http :bundle :export :max-relationships] 1000)
+          res (into #{}
+                    (mapcat #(scroll-with-limit % store
+                                                (relationships-filter records % source_type target_type)
+                                                identity-map
+                                                {:limit limit
+                                                 :sort [{"timestamp" "desc"}
+                                                        {(name %) "desc"}]}))
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj:426]: Nice one. I've missed that corner case. thanks

File: src/ctia/http/routes/crud.clj [hunk 6]
```diff
@@ -26,13 +26,9 @@
...
   [{:keys [get-store entity identity-map]}]
-  (let [get-by-id #(-> (get-store entity)
-                       (store/read-record
-                        %
-                        identity-map
-                        {}))]
+  (let [s (get-store entity)]
     (fn [ids]
-      (keep get-by-id ids))))
+      (store/read-records s ids identity-map {}))))
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/http/routes/crud.clj:31]: Yes, it does.

File: src/ctia/stores/es/crud.clj [hunk 7]
```diff
@@ -206,6 +206,35 @@ It returns the documents with full hits meta data including the real index in wh
...
+       ids :- [s/Str]
+       ident
+       es-params]
+      (sequence
+       (comp (map :_source)
+             (map coerce!)
+             (map (fn [record]
+                    (if (allow-read? record ident get-in-config)
+                      record
+                      (when *throw-access-control-error?*
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/stores/es/crud.clj:233]: my thoughts exactly. Another side of the story is that potential attackers might abuse 502 status code to identify which ids are stored in the database bypassing by that inability to read them )

File: test/ctia/bulk/routes_test.clj [hunk 8]
```diff
@@ -430,7 +434,8 @@
                (is (= (set (get-in expected-not-found [k :errors :not-found]))
                       (set (get-in res [:errors :not-found])))))
              (doseq [[_k entity-res] (dissoc get-res :tempids)]
-               (is (= [nil nil] (take 2 entity-res))
+               (is (= #{} (set/intersection (set (get delete-bulk-query _k))
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply test/ctia/bulk/routes_test.clj:437]: thanks )

File: test/ctia/bulk/core_test.clj [hunk 9]
```diff
@@ -30,8 +30,9 @@
                                  :judgement
                                  identity-singleton
                                  {:ConfigService {:get-in-config get-in-config}
+                                  ;; NOTE "unreachable" entity === unknown store
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply test/ctia/bulk/core_test.clj]: I think you are right, and it is a hack to control `:access-control-error` 
I've replaced it with mocked IStore

File: src/ctia/bundle/core.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj]: Tests reveal id is not needed in `:sort`.

I'm currently working on a more representative test case that should also cover this corner case. But for now I will remove id from sorting set.

**Review by msprunck — COMMENTED** [review]
First pass

File: src/ctia/bundle/core.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj:428]: I've tried adding "_shard_doc" into the set but I've got `"reason": "No mapping found for [_shard_doc] in order to sort on"`

File: src/ctia/bulk/core.clj [hunk 10]
```diff
@@ -99,16 +93,13 @@
...
-             (catch Exception e
-               (log/error (pr-str e)))))
-         ids)))
+   {{:keys [get-store]} :StoreService
+    :as services} :- ReadEntitiesServices]
+  (let [store (get-store entity-type)]
+    (map #(when %
+            (with-long-id % services))
+         (binding [es-crud/*throw-access-control-error?* false]
+           (store/read-records store ids auth-identity {})))))
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bulk/core.clj:102]: the behaviour can't be the same because, in the original implementation, some records reading throws an error and try/catch replaces it with nil. And read-records returns entier collection of records. 

Also, it looks like to me that the original implementation was meant to wrap with try/catch the call to `with-long-id`. WDYT?

File: src/ctia/stores/es/crud.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/stores/es/crud.clj:233]: Should it be part of that PR? I've tried to implement that already, and it requires significant changes in tests that might be seen as unrelated to the issue.

File: src/ctia/bundle/core.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (msprunck) [reply src/ctia/bundle/core.clj:428]: Sorry, I didn't realize that it already contains `source_ref` or `target_ref` as tiebreaker

[the remark this answers is not in the document]
Comment (ereteog) [reply src/ctia/bundle/core.clj:428]: The doc is for ES7, `_shard_doc` does not exists in ES5.
Recommendation for ES5, using _doc does not work indeed (probably the same thing) because _doc is unique per shard.
In tests it's hard to reproduce: add relationships with same source_ref or target_ref with a fixed timestamp and limit 1, you should hit the problem.

File: src/ctia/stores/es/crud.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/stores/es/crud.clj]: https://github.com/threatgrid/ctia/pull/1230/commits/fd8816ebd67fa8adb0a57ade6716c0e3966456e1
I think there is a better way :) WDYT?

File: src/ctia/bulk/core.clj (hunk 10 above)
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bulk/core.clj]: Now it should behave the same. Unfortunately, there is a small code duplication ( wdyt?

File: src/ctia/bundle/core.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj]: tiebreaker is important when `search_after` is used as a pagination mechanism to retrieve all the results. for `scroll-with-limit`, it is acceptable to not use it at all. The algorithm uses `search_after` only when relationships are fully collected for some `source_ref` of `target_ref` and we must jump to some page of search results that begins with another ref. There is a test case that covers this behaviour.

**Review by msprunck — COMMENTED** [review]
LGTM, thanks for all the comments, it is very useful to understand the optimization which often results in more complex code. Could you complete de description? The reference in the IROH repository is not publicly available. Could you also complete the QA section? Here is an example https://github.com/threatgrid/ctia/pull/866#qa

File: src/ctia/bulk/core.clj [hunk 11]
```diff
@@ -99,16 +93,19 @@
...
+    (map #(when %
+            (try
+              (with-long-id % services)
+              (catch Exception e
+                (log/error (pr-str e)))))
+         (binding [es-crud/*throw-access-control-error?* false]
            (try
-             (some-> (read-entity id)
-                     (with-long-id services))
+             (store/read-records store ids auth-identity {})
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bulk/core.clj:106]: Good idea! thank you

File: src/ctia/bundle/core.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (ereteog) [reply src/ctia/bundle/core.clj]: Ok great approach to overcome the limitation of `size + offset`  of ES that works when we limit the number of sources.

File: src/ctia/bundle/core.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (ereteog) [reply src/ctia/bundle/core.clj:412]: yes. 
However I share your concern about ensuring the order.
With filter + query I confirm that we have a low control, I just know that filters order matters in ES5 (I do not know for ES7 indeed) and that we ensure `query` is the last filter in a very distant implementation detail: https://github.com/threatgrid/ductile/blob/master/src/ductile/query.clj#L64.
In any case, the filtering on the ids with a regex is a very old hack and you finish to convince me to create a task to change it:
https://github.com/advthreat/iroh/issues/6528
So keep your implementation but check that we cannot hit a lucene query limit. I did not find anything, but it's better to double check.

[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/bundle/core.clj:412]: [I know only one limit](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#index-max-terms-count)
the default number is 65536. I doubt we ever want to export that many in one batch :)

**Review by ereteog — COMMENTED** [review]
that's great.
I would like to validate the QA section before merging, can you do it please?

**Comment by DeLaGuardo** [comment]
> that's great. I would like to validate the QA section before merging, can you do it please?

Thanks! sure, I'm finishing with bundles to import as the first step for QA.

File: src/ctia/stores/es/crud.clj [hunk 12]
```diff
@@ -206,6 +206,33 @@ It returns the documents with full hits meta data including the real index in wh
...
+       (comp (map :_source)
+             (map coerce!)
+             (map (fn [record]
+                    (if (allow-read? record ident get-in-config)
+                      record
+                      (let [ex (ex-info "You are not allowed to read this document"
+                                        {:type :access-control-error})]
+                        (if suppress-access-control-error?
+                          (log/error ex)
+                          (throw ex)))))))
```
[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/stores/es/crud.clj]: that looks great, but I'm not sure that we don't have tests for throwing that exception.

[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/stores/es/crud.clj]: https://github.com/threatgrid/ctia/runs/5886146558?check_suite_focus=true#step:18:8121
I remember that particular test :)
@ereteog I'm happy to remove it as part of that PR or file an issue, wdyt?

[the remark this answers is not in the document]
Comment (DeLaGuardo) [reply src/ctia/stores/es/crud.clj]: the test is much more complicated than I thought. I'll file an issue.

**Review by ereteog — APPROVED** [review]
LGTM!
